### ISSUE TITLE
query on gitlab_project_job returns no results

### ISSUE THREAD

**Issue by jDmacD** [theapsgroup/steampipe-plugin-gitlab#52]
The query returns no data. No error message given. The `project_id` is correct, queries to other tables work fine for the same id.

Plugin version:
```json
{
    "hub.steampipe.io/plugins/theapsgroup/gitlab@latest": {
      "name": "hub.steampipe.io/plugins/theapsgroup/gitlab@latest",
      "version": "0.4.0",
      "image_digest": "sha256:6aacfb9c115176d8b9b9919e8f70df790b646d24db7663cd156bbeb292463ee1",
      "binary_digest": "sha256:5a51dd4bea9469e018b0f94987b27bdca110f73d5697f9bd7e65a52aa0ad53a6",
      "binary_arch": "amd64",
      "installed_from": "us-docker.pkg.dev/steampipe/plugins/theapsgroup/gitlab:latest",
      "last_checked_date": "2023-05-17T10:28:52Z",
      "install_date": "2023-05-09T11:11:44Z",
      "imageDigest": "sha256:6aacfb9c115176d8b9b9919e8f70df790b646d24db7663cd156bbeb292463ee1",
      "installedFrom": "us-docker.pkg.dev/steampipe/plugins/theapsgroup/gitlab:latest",
      "lastCheckedDate": "2023-05-17T10:28:52Z",
      "installDate": "2023-05-09T11:11:44Z"
    }
}
```
Gitlab Version: `15.10.7-ee`

Steampipe Version:
```shell
steampipe@3a6c42c71856:/workspace$ steampipe -v    
Steampipe v0.19.4
```

Query:
```sql
select
  *
from
  gitlab_project_job
where
  project_id = 912
limit 30;
```

**Comment by pdecat** [comment]
I've just reproduced this: the query returns immediately, but the plugin continues sending requests to the Gitlab API in background:

```
2023-09-07 08:25:07.287 UTC [TRACE] steampipe-plugin-gitlab.plugin: [TRACE] 169407510621: output wg complete - send nil row (169407510621)
2023-09-07 08:25:07.287 UTC [TRACE] steampipe-plugin-gitlab.plugin: [TRACE] 169407510621: empty row on output channel - we are done
2023-09-07 08:25:07.287 UTC [INFO]  steampipe-plugin-gitlab.plugin: [INFO]  169407510621: Plugin execute complete (169407510621)

==> /home/patrick/.steampipe/logs/database-2023-09-07.log <==
2023-09-07 08:25:07.287 UTC [TRACE] hub: readPluginResult EOF error received - stop reading (0xc0000ce780) (169407510621)
2023-09-07 08:25:07.287 UTC [TRACE] hub: goFdwIterateForeignScan returned empty row - this scan complete (0xc0000ce780)
2023-09-07 08:25:07.287 UTC [TRACE] hub: AddScanMetadata for iterator 0xc0000ce780 (gitlab)
2023-09-07 08:25:07.287 UTC [TRACE] hub: hub.getConnectionPlugin for connection 'gitlab`
2023-09-07 08:25:07.287 UTC [TRACE] hub: connectionFactory getOrCreate hub.steampipe.io/plugins/theapsgroup/gitlab@latest gitlab
2023-09-07 08:25:07.287 UTC [TRACE] hub: connectionFactory get plugin: hub.steampipe.io/plugins/theapsgroup/gitlab@latest connection gitlab
2023-09-07 08:25:07.287 UTC [TRACE] hub: connectionConfig.Type is NOT legacy 'aggregator'
2023-09-07 08:25:07.287 UTC [TRACE] hub: got metadata table: gitlab_project_job cache hit: false, rows fetched 10, hydrate calls: 0
2023-09-07 08:25:07.287 UTC [TRACE] hub: update hydrate calls counter with 0
2023-09-07 08:25:07.287 UTC [TRACE] hub: goFdwEndForeignScan, iterator: 0xc0000ce780

==> /home/patrick/.steampipe/logs/plugin-2023-09-07.log <==
2023-09-07 08:25:07.644 UTC [DEBUG] steampipe-plugin-gitlab.plugin: [DEBUG] 169407510621: listProjectJobs: projectId=5313 page=3 perPage=50
2023-09-07 08:25:08.005 UTC [DEBUG] steampipe-plugin-gitlab.plugin: [DEBUG] 169407510621: listProjectJobs: projectId=5313 page=4 perPage=50
2023-09-07 08:25:08.484 UTC [DEBUG] steampipe-plugin-gitlab.plugin: [DEBUG] 169407510621: listProjectJobs: projectId=5313 page=5 perPage=50
2023-09-07 08:25:08.842 UTC [DEBUG] steampipe-plugin-gitlab.plugin: [DEBUG] 169407510621: listProjectJobs: projectId=5313 page=6 perPage=50
2023-09-07 08:25:09.222 UTC [DEBUG] steampipe-plugin-gitlab.plugin: [DEBUG] 169407510621: listProjectJobs: projectId=5313 page=7 perPage=50
```

**Comment by pdecat** [comment]
This only seems to happen when using `limit` in the SQL query.

**Comment by pdecat** [comment]
> but the plugin continues sending requests to the Gitlab API in background

Ok, this is normal because the `listProjectJobs` does not terminate if the number of returned jobs has reached the specified `limit`.

To avoid this, the loop should be ended early with:

```
		for _, job := range jobs {
			d.StreamListItem(ctx, job)

			// Context can be cancelled due to manual cancellation or the limit has been hit
			if d.RowsRemaining(ctx) == 0 {
				return nil, nil
			}
		}
```

https://github.com/turbot/steampipe-plugin-aws/blob/main/aws/table_aws_api_gateway_rest_api.go#L186C9-L189

**Comment by graza-io** [comment]
@pdecat - thanks for flagging up the missing code for terminating pagination when limit/cancellation is reached, #63 was merged into `main` to handle those cases across paginated tables.

**Comment by pdecat** [comment]
Have no clue about the other issue though, re-read the code several times, added debug log statements, items are streamed, but no result shows in the SQL response.

Also, tried upgrading the steampipe plugin SDK and gitlab client.

**Comment by graza-io** [comment]
@pdecat I've created a branch `issue-52`, would you be able to test against this and let me know if this resolves issue?

**Comment by pdecat** [comment]
Indeed, that solves it! :rocket: :tada:

**Comment by pdecat** [comment]
I was focusing on the job ID, not `project_id` :facepalm:

**Comment by graza-io** [comment]
Fix released in v0.5.1
